### DISCUSSION TITLE
Defining custom fields under errors.extensions field and use dgs codegen to retrieve the custom fields

### DISCUSSION THREAD

**Discussion by shivagithub1** [Netflix/dgs-framework#1937] [General]
Hi, 
We are using DGS framework for our new service and we are using dgs generated client to generate the JavaClient for our clients to integrate with our api. In case of errors, As mentioned in https://netflix.github.io/dgs/error-handling/, the dgs client is returning GraphQLError object from dgs library and extensions field is of TypedError  type. However we wanted to have our own errorType values and also wanted to include additional fields within extensions map. Since the errorType is of enum type, we are not able add our own values to errorType enum. 

So we are following the graphql-java spec and returning below error response using an Instrumentation class. 
```
{
    errors: [ {
         message: "exception message",
         path: "", ///Path of the field for which exception occurred
         location: "" //Array of location in the input query where the error occurred
         extensions: {
             errorType: "BAD_REQUEST|DATA_PROVIDER_ERROR|INTERNAL_API_ERROR",
             errorDetail: "exception class name",
             isRetryable: true|false
             ....ANY ADDITIONAL FIELDS THAT WE WANT TO INCLUDE IN FUTURE
         }
       }
    ]
}
```

Now we are trying to see how we can read the fields within extensions field including errorType, errorDetail, isRetryable. The dgs graphQLResponse.getErrors() method is returning  GraphQLError object from dgs library and is providing above custom fields as expected. Below is Java object toString() on GraphQLError object. 

`[GraphQLError(message=error message, path=[field1], locations=[], extensions=GraphQLErrorExtensions(errorType=UNKNOWN, errorDetail=ValidationException, origin=, debugInfo=GraphQLErrorDebugInfo(subquery=, variables={}), classification=DataFetchingException))]`

I am looking for guidance here. Below is one option that i am thinking. 

1. Define my own GraphQLError, MyApiError at client side.
2. Define message, path, location as String type and extensions as Map<String, Object> same as graphql-java
3. Get the errors using response.getJson() and de-serialize the errors to List<MyApiError> and use it in my client library code.

**Comment by mrvaruntandon** [comment]
Could share minimal code sample to replicate this behaviour? Also, please further clarify the behaviour expected?

**Reply by jamesHau** [reply]
I have the same issue.
If the server is configured such as

```java
validationErrorsException
        .getErrors()
        .getFieldErrors()
        .forEach(
            fieldError ->
                resultBuilder.error(
                    TypedGraphQLError.newBuilder()
                        .errorType(BAD_REQUEST)
                        .path(getResultPath(handlerParameters.getPath(), fieldError))
                        .message(fieldError.getCode())
                        .extensions(Map.of("detailedMessage", fieldError.getDefaultMessage()))
                        .build()));
```

The response will look like
```java
{
  "errors": [
    {
      "message": "not today",
      "path": ["a", "b"],
      "extensions": { "detailedMessage": "broken", "errorType": "BAD_REQUEST" }
    },
```

However when performing this request with the client generated by the codegen each error will be of type `com.netflix.graphql.dgs.client.GraphQLError` where the extension property is of the class `com.netflix.graphql.dgs.client.GraphQLErrorExtensions`  which looks like
`
```java
@JsonIgnoreProperties(ignoreUnknown = true)
data class GraphQLErrorExtensions(
    @JsonProperty val errorType: ErrorType? = null,
    @JsonProperty val errorDetail: String? = null,
    @JsonProperty val origin: String = "",
    @JsonProperty val debugInfo: GraphQLErrorDebugInfo = GraphQLErrorDebugInfo(),
    @JsonProperty val classification: Any = "",
)
```

This means that any custom extension you add you cannot serialize back unless you do some manual things. How should we deal with this? Can we somehow configure the deserializer to use TypedGraphQLError (provided by com.netflix.graphql.dgs:graphql-error-types) instead?

**Comment by jamesHau** [comment]
Not the prettiest but made it work using

```java
Configuration configuration = Configuration.builder()
    .mappingProvider(new JacksonMappingProvider())
    .build();

TypeRef<List<TypedGraphQLError>> typeRef = new TypeRef<>() {};

List<TypedGraphQLError> errors = JsonPath.using(configuration).parse(result.getJson()).read("$.errors", typeRef);
```
